File: front/src/app/core/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

/**
 * Service pour gérer les commentaires d'articles.
 * Permet de récupérer les commentaires d'un article et d'ajouter un commentaire.
 */
@Injectable({
  providedIn: 'root'
})
export class CommentService {
  private baseUrl = 'http://localhost:8080/api/comments';

  constructor(private http: HttpClient) {}

  /**
   * Récupère tous les commentaires associés à un article spécifique.
   * 
   * @param articleId L'identifiant de l'article pour lequel récupérer les commentaires.
   * @returns Un Observable qui émet la liste des commentaires associés à l'article.
   */
  getComments(articleId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/article/${articleId}`);
  }

  /**
   * Ajoute un nouveau commentaire à un article.
   * 
   * @param articleId L'identifiant de l'article auquel ajouter le commentaire.
   * @param comment Le contenu du commentaire à ajouter.
   * @returns Un Observable représentant la réponse de l'API après l'ajout du commentaire.
   */
  addComment(articleId: number, comment: any): Observable<any> {
    const authToken = localStorage.getItem('authToken'); // Récupère le token d'authentification
    const headers = new HttpHeaders({
      'Authorization': `Bearer ${authToken}`, // Ajoute le token à l'en-tête
      'Content-Type': 'application/json' // Spécifie que le contenu de la requête est en JSON
    });
  
    return this.http.post<any>(`${this.baseUrl}/article/${articleId}`, comment, { headers });
  }
}

File: back/src/main/java/com/openclassrooms/mddapi/services/CommentService.java
package com.openclassrooms.mddapi.services;

import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import com.openclassrooms.mddapi.dto.CommentDTO;
import com.openclassrooms.mddapi.models.Article;
import com.openclassrooms.mddapi.models.Comment;
import com.openclassrooms.mddapi.models.User;
import com.openclassrooms.mddapi.repository.ArticleRepository;
import com.openclassrooms.mddapi.repository.CommentRepository;
import com.openclassrooms.mddapi.repository.UserRepository;

/**
 * Service qui gère la logique métier des commentaires.
 * Ce service permet d'ajouter des commentaires à un article et de récupérer les commentaires associés à un article.
 */
@Service
public class CommentService {

    @Autowired
    private CommentRepository commentRepository;

    @Autowired
    private ArticleRepository articleRepository;

    @Autowired
    private UserRepository userRepository;

    /**
     * Récupère tous les commentaires d'un article spécifié par son identifiant.
     * 
     * Cette méthode permet de récupérer tous les commentaires associés à un article donné.
     * Elle retourne une liste de commentaires ou une réponse avec un message d'erreur si aucun commentaire n'est trouvé.
     * 
     * @param articleId L'identifiant de l'article pour lequel les commentaires sont récupérés.
     * @return          La réponse HTTP contenant la liste des commentaires associés à l'article.
     */
    public ResponseEntity<?> getCommentsByArticle(Integer articleId) {
        List<Comment> comments = commentRepository.findByArticleId(articleId);
        return ResponseEntity.ok(comments);
    }

    /**
     * Ajoute un commentaire à un article spécifié par son identifiant.
     * 
     * Cette méthode permet à un utilisateur authentifié d'ajouter un commentaire sur un article. Si l'utilisateur
     * ou l'article n'est pas trouvé, ou si l'utilisateur n'est pas authentifié, une erreur est retournée.
     * 
     * @param articleId   L'identifiant de l'article auquel le commentaire est associé.
     * @param commentDTO  L'objet DTO contenant les informations du commentaire (contenu, etc.).
     * @param userDetails Les détails de l'utilisateur authentifié.
     * @return            Une réponse HTTP avec un message de succès ou d'erreur.
     */
    public ResponseEntity<?> addComment(Integer articleId, CommentDTO commentDTO, UserDetails userDetails) {
        if (userDetails == null) {
            return ResponseEntity.status(401).body(Map.of("error", "Utilisateur non authentifié"));
        }

        Article article = articleRepository.findById(articleId).orElse(null);
        if (article == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Article non trouvé"));
        }

        User user = userRepository.findByUsername(userDetails.getUsername());
        if (user == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Utilisateur non trouvé"));
        }

        Comment comment = new Comment();
        comment.setContent(commentDTO.getContent());
        comment.setArticle(article);
        comment.setUser(user);

        commentRepository.save(comment);
        return ResponseEntity.ok(Map.of("message", "Commentaire ajouté avec succès"));
    }
}
